sharding/selectionBasedProvider.go

```diff
@@ -0,0 +1,96 @@
+package sharding
+
+import (
+	"encoding/binary"
+
+	"github.com/ElrondNetwork/elrond-go/hashing"
+)
+
+type validatorEntry struct {
+	startIndex     int64
+	numAppearances int64
+}
+
+// SelectionBasedProvider will handle the returning of the consensus group by simulating a reslicing of the expanded
+// eligible list. A comparison between a real reslicing and this can be found in common_test.go
+type SelectionBasedProvider struct {
+	hasher      hashing.Hasher
+	sortedSlice []*validatorEntry
+	size        int64
+}
+
+// NewSelectionBasedProvider will return a new instance of SelectionBasedProvider
+func NewSelectionBasedProvider(hasher hashing.Hasher, maxSize uint32) *SelectionBasedProvider {
+	return &SelectionBasedProvider{
+		hasher:      hasher,
+		sortedSlice: make([]*validatorEntry, 0, maxSize),
+		size:        0,
+	}
+}
+
+func (sbp *SelectionBasedProvider) addToSortedSlice(ve *validatorEntry) {
+	sbp.size += ve.numAppearances
+
+	for i := 0; i < len(sbp.sortedSlice); i++ {
+		if sbp.sortedSlice[i].startIndex >= ve.startIndex {
```

Comment:
Could be a situation when sbp.sortedSlice[i].startIndex == ve.startIndex ? If yes, should not be resolved in a different way?

Reply:
that `equals` situation shouldn't happen because indexes should be different. However I added also the `equals` check for safety, but this should never happen.

Reply:
ok